halo2_middleware/src/zal.rs

```diff
@@ -0,0 +1,233 @@
+//! Unfortunately the "Engine" name is used in bn256 for pairings.
+//! Fortunately a ZalEngine is only used in the prover (at least for now)
+//! while "pairing engine" is only used in the verifier
+//!
+//! Initialization design space
+//! ---------------------------
+//!
+//! It is recommended that ZAL backends provide:
+//! - an initialization function:
+//!   - either "fn new() -> ZalEngine" for simple libraries
+//!   - or a builder pattern for complex initializations
+//! - a shutdown function or document when it is not needed (when it's a global threadpool like Rayon for example).
+//!
+//! Backends might want to add as an option:
+//! - The number of threads (CPU)
+//! - The device(s) to run on (multi-sockets machines, multi-GPUs machines, ...)
+//! - The curve (JIT-compiled backend)
+//!
+//! Descriptors
+//! ---------------------------
+//!
+//! Descriptors enable providers to configure opaque details on data
+//! when doing repeated computations with the same input(s).
+//! For example:
+//! - Pointer(s) caching to limit data movement between CPU and GPU, FPGAs
+//! - Length of data
+//! - data in layout:
+//!    - canonical or Montgomery fields, unsaturated representation, endianness
+//!    - jacobian or projective coordinates or maybe even Twisted Edwards for faster elliptic curve additions,
+//!    - FFT: canonical or bit-reversed permuted
+//! - data out layout
+//! - Device(s) ID
+//!
+//! They are required to be Plain Old Data (Copy trait), so no custom `Drop` is required.
+//! If a specific resource is needed, it can be stored in the engine in a hashmap for example
+//! and an integer ID or a pointer can be opaquely given as a descriptor.
+
+// The ZK Accel Layer API
+// ---------------------------------------------------
+pub mod traits {
+    use halo2curves::CurveAffine;
+
+    pub trait MsmAccel<C: CurveAffine> {
+        fn msm(&self, coeffs: &[C::Scalar], base: &[C]) -> C::Curve;
+
+        // Caching API
+        // -------------------------------------------------
+        // From here we propose an extended API
+        // that allows reusing coeffs and/or the base points
+        //
+        // This is inspired by CuDNN API (Nvidia GPU)
+        // and oneDNN API (CPU, OpenCL) https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnn-ops-infer-so-opaque
+        // usage of descriptors
+        //
+        // https://github.com/oneapi-src/oneDNN/blob/master/doc/programming_model/basic_concepts.md
+        //
+        // Descriptors are opaque pointers that hold the input in a format suitable for the accelerator engine.
+        // They may be:
+        // - Input moved on accelerator device (only once for repeated calls)
+        // - Endianess conversion
+        // - Converting from Montgomery to Canonical form
+        // - Input changed from Projective to Jacobian coordinates or even to a Twisted Edwards curve.
+        // - other form of expensive preprocessing
+        type CoeffsDescriptor<'c>: Copy;
```

Comment:
We wonder how the descriptor handling will work with a GPU implementation.
Currently you have implemented a native implementation, and we see how the `Copy` requirement and the lifetimes work well with that case: the descriptors are references to the passed `coeffs` and `base`; so they have the same lifetime.  This allows having a cached implementation that doesn't copy any vector and just reuses the non-caching msm.

But we have the following questions:
- For a GPU implementation, loading the `coeffs` and `base` into the engine will copy them, so then the lifetimes seem arbitrary.  The original `coeffs` and `base` could be dropped without any problem because the GPU holds a copy.  Moreover, the lifetime restriction may complicate the code that uses `MsmAccel` because it has to keep the original `coeffs` and `base` around for the lifetime of the descriptors.
- Then we have the topic of freeing the resources associated with the descriptors.  Since the descriptors are Copy, they are limited in what they have inside.  We imagine this descriptors will have a reference to the GPU, so that they can tell the GPU to free the resources after the descriptors are dropped; how do you envision this?

I think it will help if you have a sketch of how a GPU implementation would look like.  Or in general, an implementation that ends up copying the `coeffs` and `base`.

Reply:
I forgot to revisit this. For some context the `: Copy` constraint was added following this discussion https://github.com/privacy-scaling-explorations/halo2curves/pull/107#discussion_r1459017211 when we were planning to use `&dyn Trait` and discussing whether a Box was needed or not.

Even then, I think requiring plain old data here is OK as mentioned just above and here https://github.com/privacy-scaling-explorations/halo2curves/pull/107#discussion_r1461654354

> ```
>         The most complex descriptors should be:
>         - Pointer(s) to data
>         - size of data
>         - data in layout (for example canonical or bit-reversed permuted)
>         - data out layout
>         - enum or integer ID for operation (MSM, FFT, Coset FFT)
>         - Device(s) ID
> ```
> 
> and if those are needed, maybe those data structures should live in a hashmap in &engine, and the &engine just returns a handle ID as descriptor.

Now on the questions specifically:

> For a GPU implementation, loading the coeffs and base into the engine will copy them, so then the lifetimes seem arbitrary. The original coeffs and base could be dropped without any problem because the GPU holds a copy. Moreover, the lifetime restriction may complicate the code that uses MsmAccel because it has to keep the original coeffs and base around for the lifetime of the descriptors.

We want general code that works on CPU and GPU for ease of maintenance, and also because it's not possible to test GPU code in CI, at least until Github GPU instances are available. (https://resources.github.com/devops/accelerate-your-cicd-with-arm-and-gpu-runners-in-github-actions/)

Also some GPU like Apple Metal provide unified memory and in that case the CPU memory IS the GPU memory so the lifetime are actually accurate.

> Then we have the topic of freeing the resources associated with the descriptors. Since the descriptors are Copy, they are limited in what they have inside. We imagine this descriptors will have a reference to the GPU, so that they can tell the GPU to free the resources after the descriptors are dropped; how do you envision this?

Actually I was thinking initially that Drop could be used to customize resource freeing, and at one point when we talked about `Copy` it slipped my mind.
So we really need to remove that `Copy`

Reply:
Descriptors are only used for caching.

And the very very first version of descriptors had no Copy and worked fine for caching: https://github.com/mratsim/constantine/commit/baf2b299ef8b0f6bd9d18f638b8facf3d91ff4e7

I have the impression that we're not talking about the same thing, I'll finish the branch without copy and update Constantine to ensure the no GPU no caching use-case work fine.